openquake/hazardlib/gsim/youd_etal_2002.py

```diff
@@ -0,0 +1,131 @@
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with OpenQuake. If not, see <http://www.gnu.org/licenses/>.
+
+"""
+Module
+:mod:`openquake.hazardlib.gsim.youd_etal_2002`
+exports
+:class:`YoudEtAl2002`
+"""
+import warnings
+import numpy as np
+
+from openquake.hazardlib import const, site
+from openquake.hazardlib.imt import PGD
+from openquake.hazardlib.gsim.base import GMPE, CoeffsTable
+
+
+def _compute_magnitude_term(rup, C):
+    """
+    Returns the magnitude scaling term
+    """
+    return C["c0"] + (C["c1"] * rup.mag)
+
+
+def _compute_distance_term(dists, R, C):
+    """
+    Returns the distance scaling term
+    """
+    return (C["c3"] * dists.repi) + (C["c2"] * np.log10(R))
+
+
+def _compute_site_term(sites, C):
+    """
+    Returns the distance scaling term
+    """
+    return (C["c4"] * np.log10(sites.slope)) + \
+           (C["c5"] * np.log10(sites.T_15)) + \
+           (C["c6"] * np.log10(100 - sites.F_15)) + \
+           (C["c7"] * np.log10(sites.D50_15 + 0.1))
+
+
+class YoudEtAl2002(GMPE):
+    """
+    Implements the GMPE of Youd et al. (2002) for calculating Permanent
+    ground defomation(m) from lateral spread
+
+    Youd, T. L., Hansen, C. M., & Bartlett, S. F. (2002). Revised
+    multilinear regression equations for prediction of lateral spread
+    displacement. Journal of Geotechnical and Geoenvironmental Engineering,
+    128(12), 1007-1017.
+    """
+    #: This GMPE is based on non-subduction earthquakes with M<8
+    DEFINED_FOR_TECTONIC_REGION_TYPE = const.TRT.ACTIVE_SHALLOW_CRUST
+
+    #: Supported intensity measure types are Permanent ground deformation (m)
+    #: from lateral spread
+    DEFINED_FOR_INTENSITY_MEASURE_TYPES = set([
+        PGD,
+    ])
+
+    #: Supported intensity measure component is the horizontal
+    DEFINED_FOR_INTENSITY_MEASURE_COMPONENT = const.IMC.HORIZONTAL
+
+    #: Supported standard deviation types is total.
+    DEFINED_FOR_STANDARD_DEVIATION_TYPES = set([
+        const.StdDev.TOTAL,
+    ])
+
+    #: Required site parameters
+    REQUIRES_SITES_PARAMETERS = {
+        'slope',
+        'freeface_ratio',
+        'T_15',
+        'F_15',
+        'D50_15'}
+
+    #: Required rupture parameters are magnitude (ML is used)
+    REQUIRES_RUPTURE_PARAMETERS = {'mag', 'hypo_depth'}
+
+    #: Required distance measure is epicentral distance
+    REQUIRES_DISTANCES = {'repi'}
+
+    #: GMPE not tested against independent implementation so raise
+    #: not verified warning
+    non_verified = True
+
+    def compute(self, ctx, imts, mean, sig, tau, phi):
+        for m, imt in enumerate(imts):
+
+            if ctx.hypo_depth >= 50.0:
+                ctx.repi.setflags(write=1)
+                ctx.repi[ctx.repi < 5.0] = 5.0
+
+            R = (10 ** ((0.89 * ctx.mag) - 5.64)) + ctx.repi
+
+            if ctx.freeface_ratio.all() == 0.0:
```

Comment:
I do not understand what this part of the code does. The freeface ratio is a site-specific coefficient. So depending on the value of this param for each site C and ctx.slope should be defined accordingly. However, with `ctx.freeface_ratio.all()` you check if all the sites have a value of this parameter == 0 and then apply a correction.

Reply:
Hi Marco,

PFA herewith the response to this query. Let us know, if any modifications further are to be incorporated or if we are missing something.

Thanks,
[Explanation for the following lines of the code_sept 12 2021.pdf](https://github.com/gem/oq-engine/files/7150642/Explanation.for.the.following.lines.of.the.code_sept.12.2021.pdf)

Reply:
Hello Marco,

May we know the next steps to be taken in this pull request. 

Thanks,
Prajakta